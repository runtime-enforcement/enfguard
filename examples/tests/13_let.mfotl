LET B1(x)+ = B(x) AND B(x+1) IN
LET B2(x)+ = B1(x) AND B1(x+2) IN
LET B4(x)+ = B2(x) AND B2(x+4) IN
LET B8(x)+ = B4(x) AND B4(x+8) IN
ALWAYS (FORALL x. A(x) IMPLIES B8(x))