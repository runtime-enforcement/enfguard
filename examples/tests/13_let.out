The formula
 LET B1(x:int)+ = B(add(x:int, 1)) ∧ B(x:int) IN LET B2(x:int)+ = B1(add(x:int, 2)) ∧ B1(x:int) IN LET B4(x:int)+ = B2(add(x:int, 4)) ∧ B2(x:int) IN LET B8(x:int)+ = B4(add(x:int, 8)) ∧ B4(x:int) IN □[0s,∞) (∀x:int. A(x:int) → B8(x:int))
is enforceable and types to
 LET B1٭(x:int)+ = ((B(x:int) : Cau) ∧:L (B(add(x:int, 1)) : Cau) : Cau) IN (LET B2٭(x:int)+ = ((B1٭(x:int) : Cau) ∧:L (B1٭(add(x:int, 2)) : Cau) : Cau) IN (LET B4٭(x:int)+ = ((B2٭(x:int) : Cau) ∧:L (B2٭(add(x:int, 4)) : Cau) : Cau) IN (LET B8٭(x:int)+ = ((B4٭(x:int) : Cau) ∧:L (B4٭(add(x:int, 8)) : Cau) : Cau) IN (□[0s,∞) ((∀x:int. (A(x:int) →:R (B8٭(x:int) : Cau) : Cau)) : Cau) : Cau) : Cau) : Cau) : Cau) : Cau
[Enforcer] @1 reactively commands:
Cause:
B(1), B(2), B(3), B(4), B(5), B(6), B(7), B(8), B(9), B(10), B(11), B(12), B(13), B(14), B(15), B(16)
OK.
[Enforcer] @1 nothing to do proactively.
[Enforcer] @2 reactively commands:
Cause:
B(3), B(4), B(5), B(6), B(7), B(8), B(9), B(10), B(11), B(12), B(13), B(14), B(15), B(16), B(17)
OK.
[Enforcer] @2 nothing to do proactively.
[Enforcer] @3 reactively commands:
Cause:
B(3), B(4), B(5), B(6), B(7), B(8), B(9), B(10), B(11), B(12), B(13), B(14), B(15), B(16), B(17), B(18)
OK.
